### DISCUSSION TITLE
What about being GNU Modula-3?

### DISCUSSION THREAD

**Discussion by ghost** [modula3/cm3#1128] [General]
This is old news but GNU Modula-2 was merged into GCC-13:

https://www.phoronix.com/news/Module-2-GCC-Merged

This make me wonder if it's better for Modula-3 to be just become GNU Modula-3 and merged into GCC? Being part of GCC means your code will be GPL licensed and will be copyrighted by the FSF (an unfortunately side effect though). Any parts of the code not GPL compatible will be removed and reimplemented. This will clear any problems with the current ambiguous licensing of many parts of the code, and will very likely help the adoption of the Modula-3 language.

**Comment by mikanystrom** [comment]
If you can get the FSF to consider the inclusion of the cm3 compiler in GCC, I don't know who would resist.

The libraries are still released under their original copyrights here, so it would really only affect the compiler.

The problem is, Richard Stallman was dead set against including the Modula-3 front-end way back when.  He took issue with the way the SRC folks had architected the compiler around gcc and blocked the inclusion.  This was over 20 years ago and maybe times have changed, and whoever controls GCC might be open to it today.

The issue is still there, though, that you can't just take DEC's code and re-release it under GPL.  You still need a release from whoever owns it today to do so.  Or alternatively, write a compiler front-end from scratch.

P.S. The reason RMS was against Modula-3 was because the architecture of our compiler is "too clean"---at least as I've heard the story. CM3 defines an intermediate language that is targeted by the CM3 front end and gets code-generated to it by cm3cg.  cm3cg is based on GCC, so it's GPL.  CM3's front-end (m3build/cm3) is not subject to GPL (FSF hates/hated this).  Also, the Modula-3 IL would be sufficient to implement many other programming languages, and that means you could write a proprietary front-end (for some new programming language XL)  in front of the GPL'd Modula-3 cm3cg and sell a proprietary compiler XLc that just requires you to download the GPL'd gcc+m3 compiler to use as your back-end.   Someone please fill us in if I forgot any details here.

**Reply by ghost** [reply]
The GNU Modula-2 compiler was rewritten from scratch for the inclusion into GCC. It originally generate C++ code and use g++ to produce binaries. Not very different from CM3. It was rewritten from scratch based on the Fortran frontend (gfortran) and the original compiler that generated C++ code got deprecated and will be removed.

**Reply by ghost** [reply]
Something like cm3 in it current form is unacceptable by GCC and I think they are right about that. A GCC based compiler should be able to process source code given to it directly, like this: `gcc foo.c -o foo`. The build system must be separated from the compiler. Quake must be separated from the compiler, like GNU make is separated from GCC. And in no way Quake is required to instantiate generics like what currently done.

**Reply by ghost** [reply]
Being a GCC frontend, this means use GCC's technology. There is no place for your own IL. GCC has more than enough of them already:

https://blog.aeste.my/2015/07/27/exploring-the-intermediate-representation-of-gcc/

**Reply by ghost** [reply]
Of course, everything must be in GPL. From what you presented I think you are cheating on the GPL (GPL violation). Your frontend is a GPL derived work, so it must be in GPL. Defining an intermediate language and have the GCC frontend to process input in this intermediate language instead, so you are able to keep the frontend that process Modula-3 code into this intermediate language not in GPL. It's definitely cheating. This architecture is not "too clean", it's cheating. That intermediate language must be abolished, two frontends merged into only one frontend that process Modula-3 code and use GCC's backend to generate binaries. Long story short, you are trying to be clever and cheating to take advantage of other people's work. This is a problem you created for yourselves and I think the FSF was right. I know CM3 is created long before LLVM, but LLVM is a perfect match for you if you want to play this intermediate language game.

**Reply by ghost** [reply]
For the build system, quake, it could be in any license you want. It's because you integrated the build system into the frontend (m3build/cm3) that caused a mess.

**Reply by mikanystrom** [reply]
The point is, the frontend is not GPL derived.  There is no GPL code in it.  I don't see how it can be a GPL violation in any way.

Now yes, can you say that FSF is within their rights to reject a design like this from inclusion into their system?  Well of course they are.  And that's where things have been stuck for about 25 years.

I think what really needs to be done here is to figure out who really owns the current code, get them to agree to re-release it under a Berkeley/MIT-style license, and take it from there (use the pieces to build a GPL compiler or whatever can be done, have it supported by another company possibly).  As long as the compiler is under an old DEC license, nothing will happen on that front, it's just impossible. 

Unless as you say one were to rewrite the whole thing from scratch but I think that's a tall order, especially when cm3 does exist and is good enough for many uses.

**Reply by demoitem** [reply]
Who exactly owns the IP? I thought Critical Mass bought it all from DEC or their successors Compaq. But since CM is defunct who
obtained the copyright? The creditors of the company? Or maybe the state. There is a vague concept of unknown ownership in
English law not well tested in US courts. For individuals copyright extends from the death of the creator for 70 years or so. For companies what is the law? It's all a bit murky.

**Comment by dragisha** [comment]
Not possible for different reasons, mostly licensing.

A very old story from DEC SRC times. I probably do not remember it all, but FSF demanded all source code of compiler to be GPLed and it was unacceptable for DEC. Bill Kalsow found a roundabout, m3cc was born. But then GCC started to change their IR spec… thus “when will m3cc work iwth GCC xx?”, “what with m3gdb” etc.

Best forget this. Freedom has many flavours but FSF accepts only theirs own.

**Reply by ghost** [reply]
The FSF was right. If you want to be under their umbrella, then you have to follow their rules. Defining an intermediate language and have the GCC frontend to process input in this intermediate language instead, so you are able to keep the frontend that process Modula-3 code into this intermediate language not in GPL. This is cheating. Don't try to be clever. The FSF is not stupid. I know CM3 is created long before LLVM, but LLVM is a perfect match for you if you want to play this intermediate language game.

**Comment by dragisha** [comment]
Being right is in the eye of beholder, obviously.

Companies do not have such freedom and companies like to keep right to make money out of their investments, even in ways (oh, blasphemy!!!) not approved by RMS.

IMO, this was exercise of RMS's pettyiness. Changes in IR practically killed Modula-3 potential and while FSF was fighting this particular RMS’s war, LLVM came to wide open market once dominated by GCC. LLVM opened and standardized their IR, opened all things GCC was using to shanghai developers…

In the end, RMS did good thing (irony alert!) - he enabled demonopolization of free compiler offering. In fast, stupid thing they did but it is probably only me thinking that :).

**Reply by ghost** [reply]
Stop repeat again and again about the changes in GCC's IR. This is disinformation. The particular changes happened since GCC 4.0 release and it has been stable since then. This changes in the IR spec is for good's sake. It's you that don't have the resources to continue after the changes. You have an unreasonable expectation that things should be the same forever and forever. It's GCC 13 now. They have not yet doing such a change again. It's you that shoot yourselves in the foot. If you don't create that intermediate language and rely on GCC's technology solely, the porting will be pretty much seamless and _they will help you_. Don't try to be clever and stop playing victim.

**Reply by ghost** [reply]
GNU Modula-2 is a single man project. It will not be possible without the cooperation between Gaius Mulley and the GCC developers. Without _GCC people's help_, GNU Modula-2 will not going to be a thing like it's now. The key is, be patient and don't try to be smart, cooperation and mutual help will come naturally. You don't have any of those.

**Comment by RodneyBates** [comment]
On 12/29/22 12:14, tn1997tn wrote:
>
> Of course, everything must be in GPL. From what you presented I think you are cheating on the GPL (GPL violation). Your frontend is a GPL derived work,
>
I believe nothing in the front end or the RTS is GPL licensed.  Nor is any of it
linked into the GCC-derived code generator.

**Reply by ghost** [reply]
You cheat by creating an intermediate language. People not stupid. They see your cheat.

**Reply by mikanystrom** [reply]
I think you are missing the point a bit here.  The issue is not really about Modula-3.

The GPL (of that era too) is very specific about what you can and can't do with their code without infecting your own code with GPL.  What you can't do is *link* to GPL code and then not live with the fact that your linked code becomes GPL.

SRC M3, and PM3, and CM3, don't do that.  They don't violate the GPL.  End of story.

What you have to remember is that the SRC M3 license was developed by some corporate lawyers in the mid-90s, long before corporate America had a real understanding of how to engage with free software.  As a result, the license is "oddball".  That is why I think it would be effective to get it re-released as MIT or BSD---because I think if DEC had released the code in today's world, it *would have been* under an MIT or BSD type license.

Now if we continue the hypothetical that CM3 were released under MIT or BSD license (it's not, so this is hypothetical for the moment at least), there is nothing stopping you (or the FSF, or anyone else) from re-releasing the CM3 code under the GPL.  MIT and University of California licenses are in that sense "compatible" with the GPL (and the FSF recognizes this).

But you still have a problem, and this is where RMS's posturing comes in.  The architecture of CM3 is such that a text-based IL from m3front, which is not GPL (in my hypothetical it is say MIT).  The back-end which *is* GPL takes the text-based IL and compiles it to object code.  You can now re-release GM3 where all of it is GPL.  Fine!  But.... another actor can now come along, let's say a company called Macrohard, they make a front-end for another language, Macrohard C++, and their compiler mhc++ is naturally an expensive commercial product, which produces GM3 IL as *its* output.  You've effectively allowed Macrohard to build a proprietary compiler using GCC as its back end.  That I think is what upset RMS about SRC's approach to their M3 compiler.

Of course Macrohard (and CM3) can also sidestep all this by producing C or C++ output and compiling that to object code using gcc or g++... which is the route Jay has taken with his work.  That also sidesteps the GPL, and it's really not something FSF can do anything about since they do claim to distribute a conforming C/C++ compiler.

I think you also have to understand a bit about how big American corporations work, as dragisha said.   But I don't think it's necessarily as simple as "they try to make money off their investments."  There are reasons for a company to distribute software for free in order to make money.  But also, there's a much simpler reason a lot of open-source software is produced within corporations: We who work for the corporations also have goals and motivations, and it's well understood that our personal goals and motivations are *not* entirely aligned with the corporations.  I, as an engineer, may well want to spend some time and effort to write code to give away.  

But whether it is my own personal motivation or my employer's strategic plans for some reason that makes open-source happen in my corporation, the one thing most big companies will *not* do is subject themselves to unnecessary cost or risk.  GPL is absolutely hated in the corporate world for this reason: not that it forces us to give software away (we do that all the time anyhow), but that it forces us to build a distribution infrastructure for source code that tracks all the specific versions used in specific products (we would have to know precisely what version of everything was used to build a specific product), OR ELSE, (or ALSO...) we expose ourselves to legal risk for .. .giving something away....!  Legal just hates that.  I can sit in my little silo in the company, distribute some code under GPL that no one else knows about, then the whole company gets sued for using that software in some inappropriate way... meanwhile I and everyone else that knows anything about the software, the product, or anything related, has long since moved on, new projects, maybe left the company, retired, etc etc etc.  GPL is seen as a black hole--it may be a bit exaggerated but I think there is some justification to it, since the intent of GPL is exactly to set up that sort of dynamic.  You may get a corporation that engages with GPL on a very important product, but to do it for some engineer's or some small team's side project or even main project... not gonna happen.  We'll just distribute our code MIT or BSD and let someone else repackage it under GPL if that's really important.  In my company, we distribute products that need GPL code to work without that code and let people download it from the Internet.

(Note I am stating my personal opinion here, these are not official statements of my employer.  I am just an engineer.)

**Reply by ghost** [reply]
They don't violate GPL because you found a way to cheat, the intermediate language. People is not stupid, they know you are cheating, so they are against you, and it's entirely right for them to do so.

**Reply by ghost** [reply]
From your wall of text I learned one thing, stay away from corporate junk like CM3. Goodbye.

**Comment by dragisha** [comment]
My mistake to even start talking to a guy using “cheat” and similar.

IR was there. SRC people created inerface to it, an “assembly” if you like.

Thank you for your participation. Now direct your energy into cm3.

**Reply by dmuysers** [reply]
RMS presents himself as an atheist and is said to often wear a button that reads "Impeach God", but his GPLism is a kind of (fanatic) substitute religion. Personally, as much as I am for free software, I systematically avoid to use GPLed code. Thanks goodness we have MIT and BSD.

**Comment by RodneyBates** [comment]
On 12/31/22 08:02, Mika Nyström wrote:
>
> I think you are missing the point a bit here. The issue is not really about Modula-3.
>
> The GPL (of that era too) is very specific about what you can and can't do with their code without infecting your own code with GPL. What you can't do is /link/ to GPL code and then not live with the fact that your linked code becomes GPL.
>
> SRC M3, and PM3, and CM3, don't do that. They don't violate the GPL. End of story.
>
> What you have to remember is that the SRC M3 license was developed by some corporate lawyers in the mid-90s, long before corporate America had a real understanding of how to engage with free software. As a result, the license is "oddball". That is why I think it would be effective to get it re-released as MIT or BSD---because I think if DEC had released the code in today's world, it /would have been/ under an MIT or BSD type license.
>
> Now if we continue the hypothetical that CM3 were released under MIT or BSD license (it's not, so this is hypothetical for the moment at least), there is nothing stopping you (or the FSF, or anyone else) from re-releasing the CM3 code under the GPL. MIT and University of California licenses are in that sense "compatible" with the GPL (and the FSF recognizes this).
>
> But you still have a problem, and this is where RMS's posturing comes in. The architecture of CM3 is such that a text-based IL from m3front, which is not GPL (in my hypothetical it is say MIT). The back-end which /is/ GPL takes the text-based IL and compiles it to object code. You can now re-release GM3 where all of it is GPL. Fine! But.... another actor can now come along, let's say a company called Macrohard, they make a front-end for another language, Macrohard C++, and their compiler mhc++ is naturally an expensive commercial product, which produces GM3 IL as /its/ output. You've effectively allowed Macrohard to build a proprietary compiler using GCC as its back end. That I think is what upset RMS about SRC's approach to their M3 compiler.
>

Hmm, I didn''t realize this.  So if I understand correctly, FSF's objection is not so
much about the SRC M3 front end not being GPL'd, but rather SRC's creation of cm3cg,
automatically GPL'd because it's derived from FSF's own GCC, athus providing an opening
for other, fully proprietary, front ends, using SRC's IR.  Of course, such folks could
also create their own cm3cg, it would just be a bit more work.
>
> Of course Macrohard (and CM3) can also sidestep all this by producing C or C++ output and compiling that to object code using gcc or g++... which is the route Jay has taken with his work. That also sidesteps the GPL, and it's really not something FSF can do anything about since they do claim to distribute a conforming C/C++ compiler.
>

And thus the opening was already there anyway, if with a less convenient IR.

Do I correctly understand, also, that GCC's IR is in-memory data structure, not
a sequential file?  Which to be used in such a way would have to be linked in
to GCC, thus GPL-infected.

I don't know first hand, but I have heard the criticism that FSF intentionally
avoided documenting GCC's IR, to make it more work to splice on front ends.

**Reply by mikanystrom** [reply]
Rodney, yes, this is all my current understanding.

I have a copy of RMS's statements in an old email thread somewhere but haven't been able to find them.  I am not sure he spelled out the objections in detail but I didn't see another way of interpreting them.  As I wrote in response to your other comment, DEC as author-owner of the IP can obviously release the compiler under many different licenses, and that won't affect the GPL-ness of GCC.  But the introduction of the text-based M3 IL definitely does modularize GCC in a way that limits the infectiousness of its GPL, which was SRC's goal all along---and contrary to the aims of RMS and the FSF.   You can see why that gets us to an impasse.

I don't think the GPL license specifically calls out binary linking in its text, but I know that's considered something of a red line, and I think it probably interacts with case law related to rights to make compatible code (Google v Oracle) and such things.

**Comment by RodneyBates** [comment]
On 12/29/22 07:15, Mika Nyström wrote:
>
> If you can get the FSF to consider the inclusion of the cm3 compiler in GCC, I don't know who would resist.
>
> The libraries are still released under their original copyrights here, so it would really only affect the compiler.
>

I didn't realize this.  So reimplementing without the licensing problems would entail
only doing the front end, not the RTS?

**Reply by mikanystrom** [reply]
I'm just saying that you can release whatever you want under GPL.  The Modula-3 libraries have once been released under the DEC license so they are always under that license in the state they are in right now.  You can re-release them under GPL but someone can still use them under the DEC license since they were once released under that license, and there's no clause in the license that says that if such-and-such happens, the DEC license is revoked.  The DEC license is perpetual.  So even if you somehow could re-release all of CM3 under GPL, a user of the code is not necessarily bound by that if he can point to the DEC license applying to the same code.  This is what allows the Linux folks to pull in MIT and BSD code and slap a big GPL notice on the whole thing, even though it doesn't really apply to the bits of their system that were released under the (strictly) less restrictive MIT and BSD licenses.

**Comment by dragisha** [comment]
I do not think they bought ownership of SRC Modula-3. It had a lot of various DEC work in it, bought by Compaq which was bought by (very alive) HP. They had rights to continue work, publish etc while respecting existing licenses… Just like pm3 did, or anybody doing derivative work. Those BSD guys, for example.
cm3’s stake there is code they rewrote/added. Someone may try to identify it. Personally, I do not see practical value in GPLing Modula-3. Just look how much time Gaius was pushing GNU Modula-2, and think about time and work needed.

LLVM, C backend… Maybe GNU Modula-2 backend...

**Reply by mikanystrom** [reply]
I am making another effort to identify who owns SRC Modula-3.  I will update everyone by the end of January.

I think everything you write, Dragisha, is correct.  There are two things that confound the situation just a little bit:

1. While "HP" itself is obviously alive, HP nowadays is just a trademark.  The HP that acquired Compaq has since split into at least four different companies that I can think of, and probably more.  It is likely but not certain that the title to SRC IP would have gone not to "HP" but to "HPE", which is another company.  I'm going to be following up on this.

2. There were some improvements made to SRC M3 at Critical Mass, Inc., which would not be owned by HP(E).  In order to get a clean license for CM3, we'd have to somehow figure that bit out too.  

I agree with you that there's not much point in GPL-ing CM3.  But putting CM3 under a recognizable open-source license instead of a 12-clause roll-your-own license that Digital did in the mid 90s I think would be very helpful.  As an example, if we were under a BSD or MIT license, I think Debian and other Linux distributions would accept CM3 in their package managers almost without question.  The old Digital license scares people a bit.

**Comment by ghost** [comment]
Given what you said, who will ever want to use CM3 I wonder? With the current ambiguous licensing, using CM3 is pretty much shooting yourself in the foot. I know another project in a somewhat similar situation, but at least they have some progresses albeit very slow:

https://github.com/open-watcom/open-watcom-v2/discussions/271

**Reply by mikanystrom** [reply]
Yes and no.. no, in the sense that the licensing issues don't affect people who just want to *use* CM3.  I write plenty of code under CM3 and don't have any worries about the license, for example.

But yes, redistributing the CM3 system might be another story.  Especially by any sort of commercial actor.  People get worried and have in the past made a fuss about it.  And that seems to have been one of the issues on some of the more popular platforms we've engaged with in the past (Debian).  Which makes it hard to get new users into CM3, since it's not really obvious how to install it from github---we have a high barrier to entry.  Although it has gotten a fair bit lower in the past years I think, thanks to work by Jay and others.  I have to say though, it would be really cool if you could just "apt-get install cm3" or similar.

In short I don't see any reason not to pursue a clarification of the license.  And I'm working on finding the right people.  Stay tuned...

**Reply by ghost** [reply]
Meanwhile, you should put a bit more love on BSD systems. FreeBSD, OpenBSD, NetBSD, DragonflyBSD. They will have no problems with packaging and distributing CM3. Create a port for CM3 first, lang/cm3. After you have the port, they will build and distribute packages of CM3 for you.

**Reply by ghost** [reply]
This is the logical thing to do. If you got stuck on Linux, there are other systems too. Maybe macports or homebrew for macOS. OpenIndiana also will have no problems with cm3 in their encumbered repository (has to be manually enabled by the users). These systems don't have any problems with CM3's licensing.

**Reply by dmuysers** [reply]
You didn't follow the discussion. The problem is: We do not know who actually owns the copyright, Full stop. And if you don't like M3, just leave it alone.

**Reply by ghost** [reply]
And do not forgot Windows, too. Currently you required people to do a bootstrap. This shouldn't be the case. You should bundle a MinGW distribution (CM3 needs a C++ compiler, it should be MinGW for portability, you can't package MSVC together with CM3 do you?) and provide ready to use distribution of CM3 (in the form of a ZIP or 7ZIP package that after extract and add cm3/bin and cm3/mingw64/bin to PATH, will be ready to use).

**Reply by ghost** [reply]
If you could, make a ready to use ZIP or 7ZIP package of m3swig too. Building it on Windows need to setup msys2 and other stuffs. Too cumbersome. I could do it but other people I know will just give up.

**Reply by ghost** [reply]
@dmuysers I'm the thread starter and I followed the whole discussion. The current licensing problems of CM3 will cause trouble with Linux, but not other systems. So, put more love on other systems, you could make CM3 more popular by doing so. This is my point.

**Reply by ghost** [reply]
As @mikanystrom said, it will only cause problems with distributing of CM3 itself. Applications developed by CM3 is not affected. You could develop anything you want in CM3 and sell it. This is what I understand from his comment.

**Comment by ghost** [comment]
https://www.phoronix.com/news/GCC-Java-Front-End-2023-Hopes
